src/dbm/dbm_shard.c

```diff
@@ -81,28 +81,36 @@ void dbm_shard_init(dbm_shard_t *shard) {
  * \author Ole Schuett
  ******************************************************************************/
 void dbm_shard_copy(dbm_shard_t *shard_a, const dbm_shard_t *shard_b) {
-  free(shard_a->blocks);
+  assert(shard_a != NULL && shard_b != NULL);
+
+  if (shard_a->nblocks_allocated < shard_b->nblocks) {
+    free(shard_a->blocks);
+    shard_a->blocks = malloc(shard_b->nblocks * sizeof(dbm_block_t));
+    shard_a->nblocks_allocated = shard_b->nblocks;
+  }
   shard_a->nblocks = shard_b->nblocks;
-  shard_a->nblocks_allocated = shard_b->nblocks_allocated;
-  shard_a->blocks = malloc(shard_b->nblocks_allocated * sizeof(dbm_block_t));
   assert(shard_a->blocks != NULL);
```

Comment:
Could you please move these asserts to be right after the malloc. This will make it easier for users to understand that they ran out of memory when they trigger.

Reply:
Sure, will do another PR. Since the malloc is conditional, it won't check the `blocks`, `hashtable` and `data` in any case; just saying.

Reply:
I mostly added these asserts to prevent spurious segfaults when we run out of memory. I think, it's a good pattern and I was planing to add a convention check for it. The proper way to implement that would be as a clang-tidy extension, but I simply haven't gotten around to it yet.